do pass groups individually to auth backend

splitbrain opened the pull request:
As discussed in https://github.com/splitbrain/dokuwiki/issues/3028, what
can be passed as a $filter to retrieveUsers() is ill defined right now.
Assuming that we can pass a | to fetch users from two different groups
is bound to fail with various backends.

This patch will ask for members of all given groups separately and merge
them in code. By passing a simple string (not a pattern) this should be
the most compatible implementation for the current situation.

It might be a good idea to allow an array to be passed in the grps
filter (or even in all filters?) to signal that you want to match
multiple groups. This should be discussed in the ticket above.

splitbrain committed:
do pass groups individually to auth backend

As discussed in https://github.com/splitbrain/dokuwiki/issues/3028, what
can be passed as a $filter to retrieveUsers() is ill defined right now.
Assuming that we can pass a | to fetch users from two different groups
is bound to fail with various backends.

This patch will ask for members of all given groups separately and merge
them in code. By passing a simple string (not a pattern) this should be
the most compatible implementation for the current situation.

It might be a good idea to allow an array to be passed in the grps
filter (or even in all filters?) to signal that you want to match
multiple groups. This should be discussed in the ticket above.

annda commented on syntax/table.php:
```diff
@@ -114,31 +110,31 @@ public function render($mode, Doku_Renderer $renderer, $data)
+        $rows = [];
+        foreach ($groups as $group) {
+            $users = $auth->retrieveUsers(0, -1, ['grps' => $group]);
+            foreach ($users as $user) {
+                if (!isset($rows[$user['user']])) {
+                    // prepare row of attributes + groups
+                    $rows[$user['user']] = array_merge(
```
Unfortunately this will not work with authplain, as it does not have the `user` key in user data array. I'm not sure how much sense it would make to replace the lines with something like

```suggestion
            foreach ($users as $key => $user) {
                if (!isset($rows[$user['user']])) {
                $username = isset($user['user']) ? $user['user'] : $key;
                    if (!isset($rows[$username])) {
                        // prepare row of attributes + groups
                        $rows[$username] = array_merge(
```

annda commented on syntax/table.php:
```diff
@@ -114,31 +110,31 @@ public function render($mode, Doku_Renderer $renderer, $data)
-                return $row;
-            }, $users, array_keys($users));
+                // add group membership
+                $rows[$user['user']][$group] = self::MARK;
```
Same as above: `$user['user']` might not be set.

splitbrain committed:
compensate for backends not returing user twice

As mentioned at
https://github.com/cosmocode/dokuwiki-plugin-groupmatrix/pull/1#pullrequestreview-385600433
the plain auth backend does not return the user name in the info array
again.

splitbrain commented:
I did a simpler fix. Can you have another look?

annda reviewed (changes requested):
This is better but still does not work with the group `user`. That's why I initially put the group membership flags in an additional subarray of user info.

splitbrain committed:
put groups in a subarray to avoid naming clashes

avoids clashes between user attributes and group names.

annda commented:
Looks good to me

annda committed:
Merge branch 'master' into individualgroups
